### PR TITLE
Feat/namespaces

### PR DESCRIPTION
**Pull request by rdanitz**
introduces namespaces

- introduce $DEFAULT_NAMESPACE with value "default"
- migration script for schema v3 (resolver) and v4 (core) for namespaces
- changes to core and resolver data types to accomodate namespaces
- updates to tests accordingly

### CODE REVIEW

**Comment by rdanitz** [comment]
@jerrytrieu please have a look; the migration is kindof messy due to rewiring all foreign key constraints (first disable them, then modify primary key on `Package`, ther rewire to that index).

I reckon it is necessary to change PK on `Package` from `(name, version)` to `(namespace, name, version)`, otherwise we would not get unique packages with different namespaces without violating the PK constraint. OTOH, it is sufficient to only have namespace in `Package` and resolve the namespace for all other entities by lookup -> more burden on application logic, and less on changing the schema.

Opinions?

**Comment by jerrytrieu** [comment]
We need to namespace vehicles as well. :( My mistake. Even though VINs are globally unique, they should only appear for the user account that created them.

**Comment by codders** [comment]
And I wouldn't even want to guarantee that VINs are unique in this system in the long term. We might want generic device IDs, and even with the VIN case if we have 200 customers on this system we don't want them to be blocked from using a specific (development or dummy) VIN just because anthoer customer has it.

**Comment by codders** [comment]
(And also having a potential collision between objects in users' namespaces creates an information disclosure vulnerability where our customers can poke to see how many vehicles their competitors are building, and a DoS vulnerability where that could completely block production runs by pre-allocating competitor VINs)

**Comment by jerrytrieu** [comment]
Perhaps it's better/easier to have different schemas per namespace/tenant? @rdanitz Can you look at switching schemas in slick?

**Comment by koshelev** [comment]
I don't think namespace and/or tenant are proper solution for the problem. I believe it should be solved using groups/roles/permissions/ACL.

**Comment by koshelev** [comment]
> Namespacing solves the allocation problem because users are no longer trying to access the same resources .... users can only access resources in their own namespace.

What is a cardinality between user and namespace? One-to-one? Can resource be moved to another namespace?

**Comment by codders** [comment]
For April 22nd, it's 1-to-1. For later on, I expect it will be n-to-m - users in multiple namespaces, namespaces with multiple users.

File: common-data/src/main/scala/org/genivi/sota/data/Namespace.scala [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+package org.genivi.sota.data
+
+import eu.timepit.refined.api.Refined
+import eu.timepit.refined.auto._
+import eu.timepit.refined.string._
+
+
+object Namespace {
+  // TODO schema for namespace; URI seems reasonable enough
+  type Namespace = String Refined Uri
```
Comment (rdanitz) [inline common-data/src/main/scala/org/genivi/sota/data/Namespace.scala]: URI for namespace ok? that way we cover XML namespaces, e-mails, java-like packages (although not bi-directional) etc.

**Comment by rdanitz** [comment]
@jerrytrieu ready to merge when checks have passed

File: core/src/main/scala/org/genivi/sota/core/db/Vehicles.scala [hunk 2]
```diff
@@ -75,27 +79,29 @@ object Vehicles {
...
+  def deleteById(vehicle : Vehicle) : DBIO[Int] =
+    vehicles.filter(v => v.namespace === vehicle.namespace && v.vin === vehicle.vin).delete
 
   /**
    * Find VINs that match a regular expression
    * @param reg A regular expression
    * @return A list of matching VINs
    */
-  def searchByRegex(reg:String) : Query[VehicleTable, Vehicle, Seq] = vins.filter(vins => regex(vins.vin, reg))
+  def searchByRegex(reg:String): Query[VehicleTable, Vehicle, Seq] =
```
Comment (rdanitz) [inline core/src/main/scala/org/genivi/sota/core/db/Vehicles.scala]: we search here across all namespaces; postponed until authN+authZ is implemented

**Comment by jerrytrieu** [comment]
//cc @simao  wait for this, avoid conflict

File: external-resolver/src/main/scala/org/genivi/sota/resolver/common/NamespaceExtractor.scala [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import eu.timepit.refined._
+import eu.timepit.refined.string._
+import org.genivi.sota.data.Namespace._
+
+/**
+  * Helpers for extracting namespace from request context.
+  */
+
+trait Namespaces {
+  def extractNamespace(implicit system: ActorSystem): Namespace = {
```
Comment (jerrytrieu) [inline external-resolver/src/main/scala/org/genivi/sota/resolver/common/NamespaceExtractor.scala]: May need different way to do this. Directives?

**Comment by jerrytrieu** [comment]
Looks good. @koshelev , @codders ?

**Comment by koshelev** [comment]
`extractNamespace` should be a directive

**Comment by rdanitz** [comment]
I'm already working on it: #326

**Comment by rdanitz** [comment]
obsolete

**Pull request closed without merging**